### ISSUE TITLE
Using "rawPayload" in pubsub converts payload to string

### ISSUE THREAD

**Issue by alecor191** [dapr/js-sdk#447]
## TL;DR

**Scenario**: An app sends a `UInt8Array`/`Buffer` (Protobuf encoded) payload using JS SDK to a Dapr pubsub component. A second app subscribes to the pubsub and receives the published payload (also using JS SDK). Both use Dapr with gRPC protocol and `rawPayload` set to `true`.

I'm using the Redis Docker Container created by `dapr init` as my pubsub persistent store for this example.

**Expected result**: The subscribing app's callback receives the same `UInt8Array`/`Buffer` content passed as parameter.

**Actual result**: The subscribing app's callback receives an object where the actual content is in a `data` property *as string*.

## Scenario description

I have the following simple test setup:

### Publisher App
An app publishing a message to Dapr pubsub using JS SDK. The payload is a Protobuf byte array. `rawPayload` is set to `true`:

```typescript
const dto = ... // UInt8Array containing Protobuf-encoded payload (also tried using Buffer object - see below for details
const pubSubCallSucceeded = await daprClient.pubsub.publish(..., ..., dto, { rawPayload: true });
```

Here an example content of `dto`:

```text
Buffer(99) [10, 79, 104, 101, 108, 108, 111, 32, 102, 114, 111, 109, 32, 110, 111, 100, 101, 45, 97, 112, 112, 32, 111, 110, 32, 108, 111, 99, 97, 108, 104, 111, 115, 116, 32, 119, 105, 116, 104, 32, 80, 114, 111, 116, 111, 98, 117, 102, 32, 109, 101, 115, 115, 97, 103, 101, 32, 50, 48, 50, 51, 45, 48, 49, 45, 50, 56, 84, 49, 55, 58, 53, 53, 58, 48, 55, 46, 57, 49, 55, 90, 26, 12, 8, 251, 194, 213, 158, 6, 16, 128, 143, 151, 193, 3, 34, 2, 3, 4, buffer: ArrayBuffer(8192), byteLength: 99, byteOffset: 3408, length: 99, Symbol(Symbol.toStringTag): 'Uint8Array']
```

### Subscriber
A second app that is subscribed to the Dapr pubsub also using the JS SDK. It is also configured to retrieve raw payloads:

```typescript
server.pubsub.subscribe(
  ...,
  ...,
  data => processProtobufMessage(data),
  ...,
  { rawPayload: true });
```

where `processProtobufMessage()` is defined as:

```typescript
async processProtobufMessage(payload: unknown): Promise<void> { ... }
```

What this function receives in `payload` is the following:

```text
data: '\nOhello from node-app on localhost with Protobuf message 2023-01-28T17:55:07.917Z��՞����"'
datacontenttype: 'text/plain'
id: '8e9306cd-7539-4df8-b682-db58dfb17f19'
pubsubname: 'pub-app-protobuf'
source: 'node-app'
specversion: '1.0'
time: '2023-01-28T18:55:32+01:00'
topic: 'pub-app-protobuf'
traceid: '00-00000000000000000000000000000000-0000000000000000-00'
traceparent: '00-00000000000000000000000000000000-0000000000000000-00'
```

What to note:

- The original payload is in the `payload.data` property **as string**.
- A property `datacontenttype` is set to `text/plain` 

Here I would have expected that I receive the raw payload back as `Buffer`. Is that a fair assumption or am I missing something?

### Observations in GRPCClient's `pubsub.js`

Digging deeper, it seems publishing binary payload (`Buffer`) doesn't result in the correct content type being set (I opened #448 to track that issue). That seems to explain the `text/plain` content type.

As next step I updated the GRPCClient's code to address that issue (i.e. it sets `application/octet-stream` content type when `Buffer` is provided). While now the content type was properly set, it seems the payload was still a string; just this time it was Base64 encoded.

Here what my subscribing app receives in the `processProtobufMessage()` function mentioned above:

```text
data_base64: 'Ck9oZWxsbyBmcm9tIG5vZGUtYXBwIG9uIGxvY2FsaG9zdCB3aXRoIFByb3RvYnVmIG1lc3NhZ2UgMjAyMy0wMS0yOVQxOTowMTozOC44MDNaGgwIkoXbngYQgI7pigMiAgME'
datacontenttype: 'application/octet-stream'
id: '57bdf57e-8cc1-43ae-9196-16cbc8eb0713'
pubsubname: 'pub-app-protobuf'
source: 'node-app'
specversion: '1.0'
time: '2023-01-29T20:01:59+01:00'
topic: 'pub-app-protobuf'
traceid: '00-00000000000000000000000000000000-0000000000000000-00'
traceparent: '00-00000000000000000000000000000000-0000000000000000-00'
tracestate: ''
type: 'com.dapr.event.sent'
```

### pubsub store content after publishing a message

Here comparison of the Redis dump for the pubsub component using the original JS SDK code:

```
\x0f\x01\x10\x00\x00\x01\x85\xfe\xd7\xc5O\x00\x00\x00\x00\x00\x00\x00\x00\xc3A\xa8B[\a[\x02\x00\x00\n\x00\x01\x01 \x02\t\x01\x84data\x05\x00\x01\x02 \r\x05\x00\x01\xe2:{\"@\x10\x1f\":\"\\nOhello from node-app on loc\x1falhost with Protobuf message 202\x1a3-01-29T18:43:38.244Z\\u001a@\x05\x010b`\x05\x068\\ufffd\x80\x05\x01\xda\x9e`\x13\x006@\x05\x0110\x80\x13\xe0\x03\x05\x02u\\\"@\x1a\x0102`\x05\x003`\x05\x024\",`\xae\ncontenttype \xb9\ttext/plain \x1e\x01id \x11\x1fdf14039d-633c-457a-b1e3-e9b2decf\x0111`J\bpubsubnam@E \x0c@\xf0\x01-p\xa0\xde K\x04sourc@\x1b\xc1\r@\x13\tpecversion h\x021.0 \x13\x01ti`A\xe1\x03\a\x009A\a\x0651+01:0`\"\x03opic 0\xe0\ne\x04trace`\xb6\x0200- \xf0\xe0\x14\x00\xe0\b  \x10@\x81@C\x02par!\x14 d`\x15\xe0\x15\x00 6\xe0\x05\x00 \x10\xc0G\x03stat@\xf5@\x0f\x81a\ncom.dapr.ev a\x01.s \x04\x06\"}\x04\xbc\x04\x01\xff\x01\x81\x00\x00\x01\x85\xfe\xd7\xc5O\x00\x01\x0cnode-sub-app\x81\x00\x00\x01\x85\xfe\xd7\xc5O\x00\x01\x00\x00\x01\x85\xfe\xd7\xc5O\x00\x00\x00\x00\x00\x00\x00\x00O\xc5\xd7\xfe\x85\x01\x00\x00\x01\x01\x0cnode-sub-appO\xc5\xd7\xfe\x85\x01\x00\x00\x01\x00\x00\x01\x85\xfe\xd7\xc5O\x00\x00\x00\x00\x00\x00\x00\x00\t\x00t\xb8\x0b6\x82\xcfOd
```

Notice the `contenttype \xb9\ttext/plain` and `"\\nOhello from...` (Protobuf payload starts with bytes `10`, `79`, `104`, `101` - corresponding to `\nOhe` etc.).

And here the Redis content of the pubsub message after modifying `GRPCClient`'s code to set `application-octet` content type: 

```
\x0f\x01\x10\x00\x00\x01\x85\xfe\xd5q\x81\x00\x00\x00\x00\x00\x00\x00\x00\xc3A\xceBO\aO\x02\x00\x00\n\x00\x01\x01 \x02\t\x01\x84data\x05\x00\x01\x02 \r\x05\x00\x01\xe2.{\"@\x10\x1f_base64\":\"Ck9oZWxsbyBmcm9tIG5vZG\x1fUtYXBwIG9uIGxvY2FsaG9zdCB3aXRoIF\x1fByb3RvYnVmIG1lc3NhZ2UgMjAyMy0wMS\x1f0yOVQxODo0MDo1OS44OTlaGgwIu/vang\x0fYQwPG+sAMiAgME\",`\x94\ncontenttype \x98\x17application/octet-stream ,\x01id \x1f\x1f9d7efdf4-6a1a-4bf5-b074-18fbc28e\x0173 \xdf\n,\"pubsubnam@S \x0c\x00- W\b-protobuf \x1f\x04sourc@\x1b\x03node@\x1c@\x13\x06pecvers z h\x021.0 \x13\x01ti`A\x172023-01-29T19:41:19+01:0`\"\x03opic 0\xe0\ne\x04trace`\xb6\x0300-0\xe0\x16\x00\xe0\b  \x10@\x81@C\x02par!\" d`\x15\xe0\x15\x00 6\xe0\x05\x00 \x10\xc0G\x03stat@\xf5@\x0f\x81o\ncom.dapr.ev a\x01.s \x04\x06\"}\x04\xb0\x04\x01\xff\x01\x81\x00\x00\x01\x85\xfe\xd5q\x81\x00\x01\x0cnode-sub-app\x81\x00\x00\x01\x85\xfe\xd5q\x81\x00\x01\x00\x00\x01\x85\xfe\xd5q\x81\x00\x00\x00\x00\x00\x00\x00\x00\x81q\xd5\xfe\x85\x01\x00\x00\x01\x01\x0cnode-sub-app\x81q\xd5\xfe\x85\x01\x00\x00\x01\x00\x00\x01\x85\xfe\xd5q\x81\x00\x00\x00\x00\x00\x00\x00\x00\t\x00h\xcf\xe0\x05\x9bE{\xaf
```

Note `contenttype \x98\x17application/octet-stream` and then the Base64 encoded payload: `"@\x10\x1f_base64\":\"Ck9oZWxsby...`.

I.e. it seems the binary buffer got converted to (Base64) string already as part of the publishing operation. Is there a way to not have Dapr encode binary payload?

### Potentially related ticket

I observed similar behavior (i.e. binary content getting converted to Base64 string) also with the .NET SDK (see https://github.com/dapr/dotnet-sdk/issues/1015) . I.e. if this is not a JS SDK-specific issue, but rather something  I should post to `dapr` project, please let me know.

**Comment by alecor191** [comment]
Closing this issue as I believe it will be addressed once https://github.com/dapr/dapr/issues/4541 is done.

**Comment by XavierGeerinck** [comment]
Note: see #453 as additional PR that will resolve some of the issue SDK wise.
